lib/cext/preprocess.rb

```diff
@@ -131,20 +153,268 @@ def write_array(name, leaking)
+        replacement: NOKOGIRI_DOC_NODE_CACHE_NEW
+      },
+      cast_native_for_value('node->_private', ';'),
+      cast_value_for_native('rb_node'),
+   ]
+  },
+  'xml_namespace.c' => {
+    gem: 'nokogiri',
+    patches: [
+      { match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+                                  replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+                                },
+      cast_value_for_native('ns'),
+      cast_native_for_value('node->_private', ';')
+    ]
+  },
+  'xml_reader.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io'),
+      cast_value_for_native('rb_io')
+    ]
+  },
+  'xml_sax_parser.c' => {
+    gem: 'nokogiri',
+    patches: [
+      {
+        match: 'NOKOGIRI_SAX_SELF(ctx)',
+        replacement: 'rb_tr_managed_from_handle_or_null(NOKOGIRI_SAX_SELF(ctx))'
+      },
+      {
+        match: /va_list args;[^}]*id_warning, 1, ruby_message\);/,
+        replacement: 'rb_funcall(doc, id_warning, 1, NOKOGIRI_STR_NEW2("Warning."));'
+      },
+      { match: /va_list args;[^}]*id_error, 1, ruby_message\);/,
+        replacement: 'rb_funcall(doc, id_error, 1, NOKOGIRI_STR_NEW2("Warning."));'
+      }
+    ]
+  },
+  'xml_sax_parser_context.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_value_for_native('error_list'),
+      cast_value_for_native('io'),
+      tuple_new_patch('ctxt', 'sax_handler')
+    ]
+  },
+  'html_sax_parser_context.c' => {
+    gem: 'nokogiri',
+    patches: [
+      tuple_new_patch('ctxt', 'sax_handler')
+    ]
+  },
+  'xml_syntax_error.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_native_for_value('ctx'),
+      {
+        match: 'VALUE msg',
+        replacement: 'VALUE msg[1]'
+      },
+      {
+        match: 'msg = ',
+        replacement: 'msg[0] = '
+      },
+      {
+        match: '&msg',
+        replacement: 'msg'
+      },
+    ]
+  },
+  'xml_xpath_context.c' => {
+    gem: 'nokogiri',
+    patches: [
+      cast_native_for_value('ctx'),
+      {
+        match: NOKOGIRI_DOC_RUBY_OBJECT_ORIG,
+        replacement: NOKOGIRI_DOC_RUBY_OBJECT_NEW
+      },
+      {
+        match: 'va_list args;',
+        replacement: 'va_list args; rb_raise(rb_eRuntimeError, "%s", "Exception:"); return;'
+      },
+      cast_value_for_native('xpath_handler'),
+      {
+        match: 'VALUE thing = Qnil;',
+        replacement: "VALUE thing = Qnil;\n  VALUE errors = rb_ary_new();"
+      },
+      {
+        match: 'xmlSetStructuredErrorFunc(NULL, Nokogiri_error_raise);',
+        replacement: 'xmlSetStructuredErrorFunc(rb_tr_handle_for_managed(errors), Nokogiri_error_array_pusher);'
+      },
+      {
+        match: 'if(xpath == NULL)',
+        replacement: "if (RARRAY_LEN(errors) > 0) { rb_exc_raise(rb_ary_entry(errors, 0)); }\nif(xpath == NULL)"
+      },
+      cast_native_for_value('(ctx->context->userData)', ';'),
+      { match: 'VALUE *argv', replacement: 'VALUE argv[32]' },
```

Comment:
This would also match
https://github.com/sparklemotion/nokogiri/blob/v1.8.1/ext/nokogiri/xml_xpath_context.c#L190
intended? Probably you want an extra `;`.

Reply:
Good catch. I'm amazed that compiled and worked really. :grin: